artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/WildcardAddressManager.java

```diff
@@ -221,19 +221,21 @@ private void addAddress(final SimpleString address, final Address actualAddress)
       }
    }
 
-   private synchronized void removeAndUpdateAddressMap(final Address address) throws Exception {
+   private void removeAndUpdateAddressMap(final Address address) throws Exception {
       // we only remove if there are no bindings left
       Bindings bindings = super.getBindingsForRoutingAddress(address.getAddress());
-      if (bindings == null || bindings.getBindings().size() == 0) {
-         List<Address> addresses = address.getLinkedAddresses();
-         for (Address address1 : addresses) {
-            address1.removeLinkedAddress(address);
-            Bindings linkedBindings = super.getBindingsForRoutingAddress(address1.getAddress());
-            if (linkedBindings == null || linkedBindings.getBindings().size() == 0) {
-               removeAddress(address1);
+      if (bindings == null || bindings.getBindings().isEmpty()) {
+         synchronized (this) {
+            List<Address> addresses = address.getLinkedAddresses();
```

Comment:
I'm confused here.. it seems to me that you can either move the synchronize to something specific to the address (synchronize (address)  { // the one you returned from getLinkedAddresses();


or the whole thing should be synchronized...

Are you sure this is the right thing?

Reply:
I'm also concerned with the dependency added.. can we move it to extra-tests?

Reply:
> it seems to me that you can either move the synchronize to something specific to the address (synchronize (address) { // the one you returned from getLinkedAddresses();
or the whole thing should be synchronized...

I've no idea TBH: consider that in other methods eg `addAndUpdateAddressMap` we treat changes in the 2 directions (the addresses referenced by `address::getLinkedAddresses` and `address`) as a whole/atomically, so I wasn't sure to further narrow the lock scope to something different.
Instead, I already have a little concern on the correctness as it is now, but I need to write a concurrent test to verify this, so will update this comment later.

> Are you sure this is the right thing?

Never :D 
From a performance perspective this is already a nice step beyond, because in the hot path of binding removal we're not anymore causing additions to be blocked. The ideal thing would be to NOT have any sparse linked address list, but an internal data structure on `WildcardAddressManager` (maybe a concurrent/or not BiMap or something similar?)

Reply:
> I'm also concerned with the dependency added.. can we move it to extra-tests?

Sure, I'm going to do it now and will write a test to verify that I haven't break any concurrent behaviour...if not, the `synchronize` changes are a minor if compared to the other improvements and we can address them later, given that we now have the right tool (the benchmark) to spot regressions (on CI as well ;) )

Reply:
About the `extra-tests` thing...
I'm not sure a `test` thing is the right place, because JMH benchs can (and should) use an annotation processor to create a `jar` that can be used on a CI to run benchmarks...JMH benchs are not junit tests (and shoudn't given that JUnit is not the proper env...)

Reply:
@clebertsuconic The optional jmh module has been created

Reply:
it is now in its own module, and can set the manifest main in its benchmark.jar and that module needs to be manually built and won't deploy.